### PR TITLE
Rework Accounts for fast squash, hashing state and checkpoint recovery.

### LINKED ISSUE #3562: AccountsDB is taking a long time to squash
#### Problem

AccountsDB is taking a long time to squash.

#### Proposed Solution

- [x]  AppendVec's should only contain a single Fork, therefore when a thread requests an AppendVec to write to, a new AppendVec will be allocated for that Fork if one isn't available.  AppendVec's should contain a counter of how many accounts are in the vec.

- [x] `AccountsDB::accounts_index: HashMap<Pubkey, Vec<Fork,AccountInfo>>`.  A pubkey for an account will point to a vector of all the forks its in.

- [x] `AccountsDB::roots: HashSet<Fork>`, a set of root forks.

- [x] `AccountDB::add_root(root: Fork)`, just adds the root to roots

- [x]  `AccountsDB::purge_fork(pruned_forks: &Vec<Fork>)`, remove the AppendVecs.  

- [x] `AccountsDB::load(pubkey: Pubkey, ancestors: &HashMap<Fork, usize>) -> Option<Account>`, load the max fork in the vector that is in ancestor set or the root set.

- [x] `AccountdsDB::update_index(pubkey: Pubkey, fork: Fork, account: &AccountInfo)`, store the account, but when appending to the vector, remove any entries for older roots.  We only need to store 1 oldest entry that is in the `AccountsDB:roots` set. For any removed roots, decrement the AppendVec's account counter.  If that counter is 0, free the AppendVec.  When the last AppendVec for the Fork is removed, remove the fork from roots.

AccountsDB::roots HashSet grows by one entry per slot.  4mb in 6 days.  Load should have 1 page fault to get to the vector, then a quick scan against the ancestor set (small), and the root set which might fault as well if it gets large enough.
 
tag: @sakridge @sambley  @carllin

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem
Accounts are not stored and indexed for fast squash, hashing state and checkpoint recovery.

#### Summary of Changes

* AppendVec holds StoredAccount, which has enough context to recover the index from storage
* squash is a `HashSet<u64>::insert`, see #3562
* AppendVec only holds data for a single fork
* AccountDB implements a parallel scan that sequentially reads the data stored in appendvecs, doesn't touch the index
* hash_internal_state uses the parallel scan to compute the hash of the fork

Fixes #3562

tag: @sakridge @sambley @rob-solana

### CODE REVIEW

File: runtime/src/accounts.rs [hunk 1]
```diff
@@ -233,319 +212,127 @@ impl Drop for Accounts {
...
+            |stored_account: &StoredAccount, accum: &mut Vec<(Pubkey, u64, Hash)>| {
+                accum.push((
+                    stored_account.pubkey,
+                    stored_account.write_version,
+                    hash(&serialize(&stored_account.account).unwrap()),
+                ));
+            },
+        );
+        let mut ordered_accounts: BTreeMap<Pubkey, (u64, Hash)> = BTreeMap::new();
+        accumilators
```
Comment (rob-solana) [inline runtime/src/accounts.rs]: accumulator

**Comment by aeyakovenko** [comment]
@rob-solana what does this test do?  `test_account_grow`?

**Comment by aeyakovenko** [comment]
@carllin I am seeing this unexpected failure

```

test blocktree_processor::tests::test_process_blocktree_with_two_forks ... thread 'main' panicked at 'assertion failed: self.last_root == 0 \|\| fork > self.last_root', runtime/src/accounts_index.rs:79:9
--
  | stack backtrace:

```

How can root go back in slots?

File: runtime/src/accounts_index.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            self.last_root = fork;
+        }
+        self.roots.insert(fork);
+    }
+    /// Remove the fork when the storage for the fork is freed
+    /// Accounts no longer reference this fork.
+    pub fn cleanup_dead_fork(&mut self, fork: Fork) {
+        self.roots.remove(&fork);
+    }
+}
```
Comment (rob-solana) [inline runtime/src/accounts_index.rs]: can you add tests for these?

**Comment by rob-solana** [comment]
> @rob-solana what do these tests do?
> 
> ```
> test_accountsdb_count_stores
> test_account_grow
> ```
> do they apply to this design?

not sure, have to look... @sakridge or @jackcmay know?

File: runtime/src/accounts_index.rs [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    fn test_get_empty() {
+        let key = Keypair::new();
+        let index = AccountsIndex::<bool>::default();
+        let ancestors = HashMap::new();
+        assert_eq!(index.get(&key.pubkey(), &ancestors), None);
+    }
+
+    #[test]
+    fn test_insert_no_ancestors() {
+        let key = Keypair::new();
```
Comment (rob-solana) [inline runtime/src/accounts_index.rs]: we have a Pubkey::new_rand()...

File: runtime/src/accounts.rs [hunk 4]
```diff
@@ -165,30 +142,34 @@ impl AccountStorageEntry {
...
-type AccountStorage = Vec<AccountStorageEntry>;
+type AccountStorage = Vec<Arc<AccountStorageEntry>>;
+type AccountStorageSlice = [Arc<AccountStorageEntry>];
 
 // This structure handles the load/store of the accounts
 #[derive(Default)]
 pub struct AccountsDB {
     /// Keeps tracks of index into AppendVec on a per fork basis
-    account_index: AccountIndex,
+    accounts_index: RwLock<AccountsIndex<AccountInfo>>,
```
Comment (rob-solana) [inline runtime/src/accounts.rs]: I"ve always hated the "Info" types in this module.  Can we change this to AcccountLocation, now?

File: runtime/src/accounts.rs [hunk 5]
```diff
@@ -634,82 +414,222 @@ impl AccountsDB {
...
+        if !is_root {
+            self.storage.write().unwrap().retain(|x| {
+                trace!("PURGING {} {}", x.fork_id, fork);
+                x.fork_id != fork
+            });
+        }
+    }
+
+    /// Store the account update.
+    pub fn store(&self, fork_id: Fork, accounts: &[(&Pubkey, &Account)]) {
```
Comment (rob-solana) [inline runtime/src/accounts.rs]: this need to be pub?

File: runtime/src/accounts.rs [hunk 6]
```diff
@@ -739,65 +659,66 @@ impl Accounts {
...
     /// Slow because lock is held for 1 operation instead of many
     pub fn store_slow(&self, fork: Fork, pubkey: &Pubkey, account: &Account) {
-        self.accounts_db.store(fork, pubkey, account);
+        self.accounts_db.store(fork, &[(pubkey, account)]);
     }
 
     fn lock_account(
-        fork: Fork,
-        account_locks: &mut HashMap<Fork, HashSet<Pubkey>>,
+        locks: &mut HashSet<Pubkey>,
```
Comment (rob-solana) [inline runtime/src/accounts.rs]: how does this work if replay stage and banking_stage are going after the same account on different slots?

Comment (rob-solana) [reply runtime/src/accounts.rs]: nevermind, bank.accounts is a new struct every time

**Pull request merged**